crates/doc/src/combine/mod.rs

```diff
@@ -0,0 +1,266 @@
+use crate::{
+    reduce,
+    validation::{FailedValidation, Validator},
+    ArchivedNode, HeapDoc, LazyNode, Pointer,
+};
+use std::io::{self, Seek};
+use std::rc::Rc;
+
+#[derive(thiserror::Error, Debug, serde::Serialize)]
+pub enum Error {
+    #[error("failed to combine documents having shared key")]
+    Reduction(#[from] reduce::Error),
+    #[error("document is invalid: {0:#}")]
+    FailedValidation(#[source] FailedValidation),
+    #[error("asked to left-combine, but right-hand document is already fully reduced: {0}")]
+    AlreadyFullyReduced(serde_json::Value),
+    #[error(transparent)]
+    SchemaError(#[from] json::schema::index::Error),
+    #[error("spill file IO error")]
+    #[serde(serialize_with = "serialize_as_display")]
+    SpillIO(#[from] io::Error),
+}
+
+pub mod memtable;
+pub use memtable::{MemDrainer, MemTable};
+
+pub mod spill;
+pub use spill::{SpillDrainer, SpillWriter};
+
+/// Accumulator is a MemTable paired with a File-backed SpillWriter.
+/// As the caller utilizes the MemTable the Accumulator will transparently
+/// spill table contents to its SpillWriter and then re-initializes a new and
+/// empty MemTable, bounding overall memory usage.
+pub struct Accumulator {
+    memtable: Option<MemTable>,
+    spill: SpillWriter<std::fs::File>,
+}
+
+impl Accumulator {
```

Comment:
I like how the Combiner got broken down into two types that encode the state transitions :+1: